specs-derive/src/lib.rs

```diff
@@ -6,61 +6,59 @@
 extern crate proc_macro;
 #[macro_use]
 extern crate quote;
+#[macro_use]
 extern crate syn;
 
 use proc_macro::TokenStream;
 use quote::Tokens;
-use syn::{Ident, MacroInput, MetaItem, NestedMetaItem};
+use syn::{Path, DeriveInput};
+use syn::synom::Synom;
 
 /// Custom derive macro for the `Component` trait.
 ///
 /// ## Example
 ///
 /// ```rust,ignore
-/// #[derive(Component, Debug)]
-/// struct Pos(f32, f32, f32);
-/// ```
-///
-/// The macro will store components in `DenseVecStorage`s by default. To specify
-/// a different storage type, you may use the `#[component]` attribute.
+/// use specs::storage::VecStorage;
 ///
-/// ```rust,ignore
 /// #[derive(Component, Debug)]
-/// #[component(HashMapStorage)]
+/// #[storage(VecStorage)]
 /// struct Pos(f32, f32, f32);
 /// ```
-#[proc_macro_derive(Component, attributes(component))]
+#[proc_macro_derive(Component, attributes(storage))]
 pub fn component(input: TokenStream) -> TokenStream {
-    let s = input.to_string();
-    let ast = syn::parse_derive_input(&s).unwrap();
+    let ast = syn::parse(input).unwrap();
     let gen = impl_component(&ast);
-    gen.parse().unwrap()
+    gen.into()
 }
 
-fn impl_component(ast: &MacroInput) -> Tokens {
+struct StorageAttribute {
+    storage: Path,
+}
+
+impl Synom for StorageAttribute {
+    named!(parse -> Self, map!(
+        parens!(syn!(Path)),
+        |(_, storage)| StorageAttribute { storage }
+    ));
+}
+
+fn impl_component(ast: &DeriveInput) -> Tokens {
     let name = &ast.ident;
     let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
 
     let storage = ast.attrs
         .first()
-        .and_then(|attr| match attr.value {
-            MetaItem::List(ref ident, ref items) if ident == "component" => items.first(),
-            _ => None,
-        })
-        .and_then(|attr| match *attr {
-            NestedMetaItem::MetaItem(ref item) => Some(item),
-            _ => None,
-        })
-        .and_then(|attr| match *attr {
-            MetaItem::Word(ref ident) => Some(ident),
-            _ => None,
-        })
-        .cloned()
-        .unwrap_or(Ident::new("::specs::storage::DenseVecStorage"));
+        .map(|attr| attr.tts.clone())
+        //.unwrap_or_else(|| quote!("(DenseVecStorage)").into());
```

Comment:
@dtolnay What's the recommended way for providing a default path here?

Reply:
Maybe something like this:

```rust
let storage = ast.attrs.iter()
    .find(|attr| attr.path.segments[0].ident == "storage")
    .map(|attr| syn::parse2::<StorageAttribute>(attr.tts.clone()).unwrap().storage)
    .unwrap_or_else(|| parse_quote!(DenseVecStorage));
```

Reply:
Thank you!